### PR TITLE
use upsert for UCR tables

### PR DESCRIPTION
**Pull request by snopoke**
Seems [COMMCAREHQ-EBZ](https://sentry.io/share/issue/97c43c6ac5c7449e9c60c4090e6ea748/) didn't get fixed by https://github.com/dimagi/commcare-hq/pull/22123.

Changing to use upsert should do the trick. Somewhat curious how this is happening in the first place. @emord is it possible that multiple pillows are processing the same kafka partition?

### CODE REVIEW

**Comment by emord** [comment]
Does this work correctly if the partitioning key changes, but the primary key does not? So for example in [this UCR](https://github.com/dimagi/commcare-hq/blob/master/custom/icds_reports/ucr/data_sources/ccs_record_cases_monthly_v2.json) the primary key is (doc_id, repeat_iteration) and the data is partitioned on month. When processed today this would look like

doc_id_1, 0, 2018-09-01
doc_id_1, 1, 2018-10-01
doc_id_1, 2, 2018-11-01

When procssessed next month, it will look like

doc_id_1, 0, 2018-10-01
doc_id_1, 1, 2018-11-01
doc_id_1, 2, 2018-12-01

My concern is that if we use upsert, the record in the 2018-09-01 partition would not be deleted by an upsert.

At least I think that's what I meant by my comment https://github.com/dimagi/commcare-hq/pull/21505#issuecomment-412164773

> is it possible that multiple pillows are processing the same kafka partition?

possible, maybe worth checking the app processes yml? I think it would be more likely that the same case is in multiple partitions because we are using a round robin partitioning approach.

I've been thinking about using the form/case ids as the partitioning key so that wouldn't happen https://kafka-python.readthedocs.io/en/master/apidoc/KafkaProducer.html#kafka.KafkaProducer.send thoughts?

File: corehq/apps/userreports/sql/adapter.py [hunk 1]
```diff
@@ -166,10 +164,19 @@ def save_rows(self, rows):
         #   the plain INSERT INTO VALUES statement resulting from below line
         #   because bulk_insert_mappings is meant for multi-table insertion
         #   so it has overhead of format conversions and multiple statements
-        insert = table.insert().values(formatted_rows)
+
+        # https://docs.sqlalchemy.org/en/latest/dialects/postgresql.html?highlight=conflict#insert-on-conflict-upsert
```
Comment (millerdev) [inline corehq/apps/userreports/sql/adapter.py]: ❤️ sqlalchemy

**Review by millerdev — APPROVED** [review]

**Comment by snopoke** [comment]
Classic, didn't realize you'd tried this already. What if just disable it for partitioned UCRs?

Re using doc ID as kafka hash key I do think that makes sense.

**Comment by emord** [comment]
> What if just disable it for partitioned UCRs?

I think most UCRs on ICDS are partitioned, so I'm not sure how helpful that would be. Seems fine if you think its worth keeping two versions of insert implementation for prod

**Pull request closed without merging**